### DISCUSSION TITLE
AWS Iot on port 443

### DISCUSSION THREAD

**Discussion by silentrain23** [arendst/Tasmota#17632] [Q&A] [answered]
Hi all,
I am trying connecting my tasmota device on a Shelly3EM to my MQTT AWS IOT account (which works fine with other ESP32 devices with my firmwares that use cert/private/CA keys and publish to arn endpoint).
I compiled with 
'#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
#define USE_MQTT_TLS_CA_CERT // Optional but highly recommended
#endif
#ifndef USE_MQTT_AWS_IOT_LIGHT
#define USE_MQTT_AWS_IOT_LIGHT
#endif
#ifdef USE_DISCOVERY
#undef USE_DISCOVERY
#endif'
in 'user_config_override.h' and I followed the cloud formation described in the 'https://tasmota.github.io/docs/AWS-IoT/' guide.

Do I have to specify the private/certs/CA cerificates anywhere? How do I check the TLS operationals in the console? The only 'ERR' i got is 'MQT: TLS connection error: 0
13:27:27.010 MQT: Connect failed to xxxxxxxxxxxx-ats.iot.eu-west-1.amazonaws.com:443, rc -4. Retry in 120 sec'

I have to say that I cannot connect to the AWS ARN with port 443 and user/password either from a different client (mqtt.fx).
Any help?

**Comment by s-hadinger** [comment]
You first need to choose whether you want password Auth or certificate. `USE_MQTT_AWS_IOT_LIGHT` is the default and uses password authentication. Also keep in mind that you need an aws iot authrnticator for it to work. 

Using port 443 requires to set ALPN to `mqtt` which is maybe why you can't connect from other clients.

**Reply by silentrain23** [reply]
I am trying going without certificates, just with credentials. I have a authorizer TasmotaAuth. For ALPN I cannot find any setting in both mqtt.fx and Mqtt Explorer clients.

**Reply by s-hadinger** [reply]
ALPN are TLS settings. They are not always exposed

**Reply by silentrain23** [reply]
@s-hadinger do you think is fine to connect to AWS IOT from tasmota with these settings?

**Reply by s-hadinger** [reply]
Yes. This is what I do for my home

**Reply by silentrain23** [reply]
do you set the 'x-amzn-mqtt-ca' anywhere for ALPN?

**Reply by s-hadinger** [reply]
No. ALPN is an additional information when establishing the TLS connection. It has nothing to do with http headers

**Reply by silentrain23** [reply]
I am really getting crazy with these settings. I tried in postman to test the authorizer that comes with tasmota cloud formation, but I got a forbidden message. Can you please send me a postman that works for you? Thank you so much?
(I am thinking to switch to cert/private/CA certificates: is this doc the best guide available? `https://tasmota.github.io/docs/AWS-IoT-cert/`)
I succeeded in compiling with TLS, did the cloud formation and pasted TLSKey1 and TLSKey2, changed endpoint and port, but nothing seem publishing and all the messages in console are RSL and not MQT (like I get if I use a broker.hivemq.com 1883 connection) these are the errors
`MQT: TLS connection error: 0
20:51:36.528 MQT: Connect failed to aoyefaXXXXXXX-ats.iot.eu-central-1.amazonaws.com:8883, rc -2. Retry in 40 sec
20:51:53.481 CMD: TLSKey
20:51:53.486 RSL: RESULT = {"TLSKey1":-1,"TLSKey2":-1}
20:52:08.679 CMD: TLSKey1 gq2....XXXX=
20:52:08.742 RSL: RESULT = {"TLSKey1":32,"TLSKey2":-1}
20:52:17.530 MQT: Attempting connection...
20:52:18.725 MQT: TLS connection error: 0
20:52:18.727 MQT: Connect failed to aoyefaXXXXXXX-ats.iot.eu-central-1.amazonaws.com:8883, rc -2. Retry in 50 sec
20:52:26.002 CMD: TLSKey2 MIICfTCC...XXXX=
20:52:26.138 RSL: RESULT = {"TLSKey1":32,"TLSKey2":641}`

At the end I succedded: I had to reboot many times, and recompile (not changing anything) because I lost the energy reading of my device. Now it publishes succesfully to AWS endpoint with both private and cert

**Comment by geri-m** [comment]
Hello, I do have exactly the same problem. I use the CF script from the [AWS Guide](https://tasmota.github.io/docs/AWS-IoT/) using username/password Authentication on port 443 (MQTT Protocol) 

That the config for compile the firmware. 

```
#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
#define USE_MQTT_TLS_CA_CERT // Optional but highly recommended
#endif

#ifndef USE_MQTT_AWS_IOT_LIGHT
#define USE_MQTT_AWS_IOT_LIGHT
#endif

#ifdef USE_DISCOVERY
#undef USE_DISCOVERY
#endif

#ifdef USE_DOMOTICZ
#undef USE_DOMOTICZ
#endif

#ifdef USE_TASMOTA_DISCOVERY
#undef USE_TASMOTA_DISCOVERY
#endif

#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif

#ifndef USE_SML_M
#define USE_SML_M
#endif

#ifdef USE_RULES
#undef USE_RULES
#endif
```

and that's the log output I'm getting

```
MQT: TLS connection error: 0
MQT: Connect failed to xxxxxxxxxxxx-ats.iot.eu-central-1.amazonaws.com:443, rc -4. Retry in 120 sec'
```

Connection to the IoT Core via MQTT Explorer or MQTT CLI is working via Websocket, but not via MQTT. 

Hence, also my assumption is, that there is a problem with ALPN on the Tasmota side. So i was wondering how I would be able to figure this out.

**Reply by Jason2866** [reply]
Using port 443 requires to set `ALPN` to `mqtt` which is maybe why you can't connect from other clients.

**Reply by geri-m** [reply]
Right @Jason2866 , that's my assumption - question is: how do i do this in Tasmota? It is actually already [part of the code](https://github.com/arendst/Tasmota/blob/ef4138bdaabad976eba1a2c29f981ab96e38284b/tasmota/tasmota_xdrv_driver/xdrv_02_9_mqtt.ino#L238), when using `USE_MQTT_TLS` (which I do) so I wonder why this is not happening.

**Reply by s-hadinger** [reply]
ALPN is automatically set when tasmota detects that the target is AWS IOT Core. What device are you using?

**Reply by geri-m** [reply]
I'm using the "Hichi IR" for reading the data from a smartmeter. What is a bit unclear to me, [AWS docs](https://docs.aws.amazon.com/iot/latest/developerguide/protocols.html) say `On port 8443 HTTPS and port 443 MQTT with ALPN x-amzn-mqtt-ca, [custom authentication](https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html) can't be used.` -- But I assume, if ALPN is set to `mqtt` only would allow custom authentication using username and pwd.

**Reply by s-hadinger** [reply]
It all depends how you want to authenticate. On port 443, you need to set:
- ALPN to `mqtt` if you use password authentication with an AWS Lambda function to check the authentication. This is the mode described in Tasmota
- ALPN to `x-amzn-mqtt-ca` if you want to use a per-device certificate authentication. This mode is not supported in Tasmota. Use instead port 8883 with no ALPN

**Reply by geri-m** [reply]
I did a bit of debugging. With the following result
   * TLS is enabled (just making sure)
   * The Error (-4) is generated by the method call `MqttClient.connect(...)` -- Error is thrown then in the PubSub implementation. 
   * In AWS Cloud Watch log I don't see a connection attempt in the AWS IoT Core Logs (and subsequently there are no calls to Authorizer. 
   * I was able to connect to a local MQTT broker

Connection to AWS works with `mosquitto_sub`. So I guess AWS is fine. Any other ideas, what I can try, what options to compile into the firmware? Can I somehow `ping` my endpoint?

**Reply by s-hadinger** [reply]
Can you PM me on Discord?
s-hadinger#8432

**Comment by Jason2866** [comment]
@s-hadinger can you help? Have you tested since IPv6 support was added?

**Reply by s-hadinger** [reply]
Yes, I have a ESP32 device connected to AWS IoT with an IPv6 address.

**Comment by s-hadinger** [comment, marked as answer]
It should be fixed by https://github.com/arendst/Tasmota/pull/17936

The SNI of TLS was broken on ESP8266 since https://github.com/arendst/Tasmota/commit/ce0a0d05a9b493aceb46bf0a7f5d46f684ad0365

**Reply by geri-m** [reply]
ack. and thanks for the quick reply.
